### ISSUE TITLE
Calibrated Confidence Scores

### ISSUE THREAD

**Issue by dantreiman** [ludwig-ai/ludwig#1908]
A calibrated confidence score, in short, means the predicted class probability should approximate the likelihood that the model is correct.  i.e. When the model predicts 0.9, it should be right about 90% of the time.

DNNs typically output high softmax probability for any input.  Softmax outputs are not generally well-calibrated confidence scores.

However, they can be calibrated for many datasets by simply scaling the logits by a temperature T (tempered softmax).  T can be chosen after training by minimizing NLL.

So, we could add a post-processing phase for binary and categorical output features which learns a value for T which minimizes NLL, as in https://arxiv.org/abs/1706.04599

This learned T could be stored in the model metadata and used to generate calibrated confidence scores at predict time.

**Comment by drishi** [comment]
+1

**Comment by tgaddair** [comment]
Closed by #1949.
